presto-main/src/main/java/io/prestosql/execution/SqlQueryExecution.java

```diff
@@ -457,6 +466,48 @@ private PlanRoot doPlanQuery()
         return new PlanRoot(fragmentedPlan, !explainAnalyze);
     }
 
+    private void sanityCheckColumnsOutput(Plan plan)
+    {
+        long count = countOutputElements(plan);
+        if (count > getQueryMaxColumnsOutput(stateMachine.getSession())) {
+            throw new PrestoException(NOT_SUPPORTED, "Too many columns output");
+        }
+    }
+
+    private long countOutputElements(Plan plan)
+    {
+        PlanNode root = plan.getRoot();
+        TypeProvider typeProvider = plan.getTypes();
+        List<Symbol> outputSymbols = root.getOutputSymbols();
+        long count = 0;
+        for (Symbol outputSymbol : outputSymbols) {
+            Type type = typeProvider.get(outputSymbol);
+            count += countFields(type);
+        }
+        return count;
+    }
+
+    private long countFields(Type type)
+    {
+        if (type instanceof ArrayType) {
+            ArrayType arrayType = (ArrayType) type;
+            return countFields(arrayType.getElementType());
+        }
+        else if (type instanceof RowType) {
+            long count = 0;
+            RowType rowType = (RowType) type;
+            for (RowType.Field field : rowType.getFields()) {
+                count += countFields(field.getType());
+            }
+            return count;
+        }
+        else if (type instanceof MapType) {
+            MapType mapType = (MapType) type;
+            return countFields(mapType.getKeyType()) + countFields(mapType.getValueType());
+        }
+        return 1;
```

Comment:
this may miscalculate any types added by the plugins, not sure if how common this usecase is. We could use `Type#getTypeParameters` recursively that solves that issue and simplify things.